ion/cmp_test.go

```diff
@@ -0,0 +1,205 @@
+	switch val := thisValue.(type) {
+	case string: // null.decimal
+		return strNullTypeCmp(val, otherValue)
+	case *Decimal:
+		thisDecimal := ionDecimal{val}
+		return thisDecimal.eq(ionDecimal{otherValue.(*Decimal)})
+	default:
+		return false
+	}
+}
+
+func cmpTimestamps(thisValue, otherValue interface{}) bool {
+	if !haveSameTypes(thisValue, otherValue) {
+		return false
+	}
+
+	switch val := thisValue.(type) {
+	case string: // null.timestamp
+		return strNullTypeCmp(val, otherValue)
+	case time.Time:
+		thisTimestamp := ionTimestamp{val}
+		return thisTimestamp.eq(ionTimestamp{otherValue.(time.Time)})
+	default:
+		return false
+	}
+}
+
+func cmpValueSlices(thisValues, otherValues []interface{}) bool {
+	if len(thisValues) == 0 && len(otherValues) == 0 {
+		return true
+	}
+
+	if len(thisValues) != len(otherValues) {
+		return false
+	}
+
+	res := false
+	for idx, this := range thisValues {
+		other := otherValues[idx]
+
+		if !haveSameTypes(this, other) {
+			return false
+		}
+
+		thisItem := getContainersType(this)
+		otherItem := getContainersType(other)
+		res = containersEquality(thisItem, otherItem)
+
+		if !res {
+			return false
+		}
+	}
+	return res
+}
+
+func cmpStruct(thisValues, otherValues []interface{}) bool {
+	if len(thisValues) == 0 && len(otherValues) == 0 {
+		return true
+	}
+
+	if len(thisValues) != len(otherValues) {
+		return false
+	}
+
+	var res bool
+	var checked []int
+	for _, this := range thisValues {
+		res = false
+		var thisItem = getContainersType(this)
+		for i := 0; i < len(otherValues); i++ {
+			if contains(checked, i) {
+				continue
+			}
+			if !haveSameTypes(this, otherValues[i]) {
+				continue
+			} else {
+				otherItem := getContainersType(otherValues[i])
+				res = containersEquality(thisItem, otherItem)
+				if res {
+					if !contains(checked, i) {
+						checked = append(checked, i)
+					}
+					break
+				}
+			}
+		}
+	}
+	if len(otherValues) != len(checked) {
+		return false
+	}
+
+	return res
+}
+
+func strNullTypeCmp(this, other interface{}) bool {
+	thisStr, thisOk := this.(string)
+	otherStr, otherOk := other.(string)
+	if thisOk && otherOk {
+		return cmp.Equal(thisStr, otherStr)
+	}
+	return false
+}
+
+func haveSameTypes(this, other interface{}) bool {
+	return reflect.TypeOf(this) == reflect.TypeOf(other)
+}
+
+func getContainersType(in interface{}) interface{} {
+	switch in.(type) {
+	case string:
+		return in.(string)
+	default:
+		return in.(ionItem)
+	}
+}
+
+func contains(list []int, idx int) bool {
+	for _, num := range list {
+		if num == idx {
+			return true
+		}
+	}
+	return false
+}
+
+// non-null containers have ionItems inside them
+func containersEquality(this, other interface{}) bool {
+	switch this.(type) {
+	case string: // null.list, null.sexp, null.struct
+		if strNullTypeCmp(this, other) {
+			return true
+		}
+	default:
+		otherItem := other.(ionItem)
+		thisItem := this.(ionItem)
+		if thisItem.fieldName == otherItem.fieldName && thisItem.equal(otherItem) {
```

Comment:
```thisItem.fieldName == otherItem.fieldName``` can only be done when both values are converted to ```ionItem``` so I put this here rather than inside ```cmpStruct```.
This still can avoid execution of ```.equal()``` if inferred as false